### PR TITLE
feat: implement user service (#3)

### PR DESCRIPTION
**Pull request by maritza05**
Implementation of the user service.

### CODE REVIEW

File: internal/service/user.go [hunk 1]
```diff
@@ -20,13 +22,19 @@ func NewUser(repo repo.UserRepository) UserService {
 }
 
 func (s UserService) Add(user entity.User) (int, error) {
```
Comment (maritza05) [inline internal/service/user.go:24]: Remove int return value.

File: internal/service/user.go [hunk 2]
```diff
@@ -20,13 +22,19 @@ func NewUser(repo repo.UserRepository) UserService {
...
-	// TODO: Implement me!
-	return 0, nil
+	err := s.repo.Create(user)
```
Comment (maritza05) [inline internal/service/user.go:25]: Add validation and create a new error for validation.

File: internal/service/user_test.go [hunk 3]
```diff
@@ -1 +1,281 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			mockRepo := mocks.NewUserRepository(t)
+			s := NewUser(mockRepo)
+			mockRepo.On("Create", user).Return(tt.repoError)
+
+			_, gotErr := s.Add(user)
+
+			if tt.wantsErr {
+				assert.Error(t, gotErr)
+				assert.Contains(t, gotErr.Error(), tt.wantErrMessage)
```
Comment (maritza05) [inline internal/service/user_test.go:55]: Validate that I got the same error.

File: internal/service/user_test.go [hunk 4]
```diff
@@ -1 +1,281 @@
...
+			wantsErr:       true,
+			wantErrMessage: "Couldn't create user",
+			repoError:      errors.New("Couldn't write to db"),
+		},
+	}
+
+	for _, tt := range testsCases {
+		t.Run(tt.name, func(t *testing.T) {
+			mockRepo := mocks.NewUserRepository(t)
+			s := NewUser(mockRepo)
```
Comment (maritza05) [inline internal/service/user_test.go:48]: Change the new method to be newUserService

File: internal/service/user.go [hunk 5]
```diff
@@ -3,14 +3,15 @@ package service
 import (
 	"github.com/wizeline/CA-Microservices-Go/internal/entity"
 	"github.com/wizeline/CA-Microservices-Go/internal/logger"
+	"slices"
 )
 
 type UserRepo interface {
 	Create(user entity.User) error
-	Read(id int) (entity.User, error)
+	Read(id uint64) (entity.User, error)
```
Comment (maritza05) [inline internal/service/user.go:11]: Changed the repo's interface to prevent lost of information

File: internal/service/error.go [hunk 6]
```diff
@@ -13,3 +16,21 @@ func (e Err) Error() string {
...
+
+type InvalidFilter struct {
+	Filter string
+}
+
+func (e InvalidFilter) Error() string {
+	return fmt.Sprintf("Invalid filter for search: %s", e.Filter)
+}
+
+var InvalidPassword = errors.New("Can't login, passwords doesn't match")
```
Comment (marcos-wz) [inline internal/service/error.go:36]: This is a small fix. The block of variable definitions use to be at the top of the file after the library import section.

File: internal/service/user.go [hunk 7]
```diff
@@ -3,14 +3,15 @@ package service
 import (
 	"github.com/wizeline/CA-Microservices-Go/internal/entity"
 	"github.com/wizeline/CA-Microservices-Go/internal/logger"
+	"slices"
```
Comment (marcos-wz) [inline internal/service/user.go:6]: Small fix. In Go is a good practice follow an importing order:  standard libraries, our libraries, third-party libraries

File: internal/service/user.go [hunk 8]
```diff
@@ -24,56 +25,148 @@ func NewUserService(repo UserRepo, l logger.ZeroLog) UserService {
...
 }
 
 func (s UserService) Get(id uint64) (entity.User, error) {
-	// TODO: Implement me!
-	return entity.User{}, nil
+	user, err := s.repo.Read(id)
+	if err != nil {
+		return entity.User{}, err
+	}
+	return user, nil
```
Comment (marcos-wz) [inline internal/service/user.go:40]: We could go straight forward and return from `s.repo.Read` function as the responses are the same. It allows reduce the code to a one-line function `return s.repo.Read(id)`

**Review by marcos-wz — CHANGES_REQUESTED** [review]
LGTM. Just minor changes

File: internal/service/user.go [hunk 9]
```diff
@@ -24,56 +25,148 @@ func NewUserService(repo UserRepo, l logger.ZeroLog) UserService {
...
 }
 
 func (s UserService) ValidateLogin(username string, password string) (entity.User, error) {
-	// TODO: Implement me!
-	return entity.User{}, nil
+	users, err := s.Find("Username", username)
+	if err != nil {
+		return entity.User{}, err
+	}
+	if users[0].Passwd == password {
```
Comment (marcos-wz) [inline internal/service/user.go:139]: We can't validate a plain password vs a hashed password. I suspect we should implement a ValidateLogin function in the repository implementation or add more logic in the service function

**Review by marcos-wz — CHANGES_REQUESTED** [review]
Integration issues

**Review by marcos-wz — APPROVED** [review]

**Pull request merged**